File: frontend/src/api/adminEval.ts
// 行政端行政分審核 API 層(/admin/eval,權限鍵 aeval):
// 分數計算/調整套用皆在後端,前端只讀 FinalScore;調整留痕於 eval_adjustments(revert=註銷不硬刪)
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import dayjs from 'dayjs'
import { api, apiPaged, qs } from './client'
import { toFinalScores, type AdScoreOut } from './eval'
import type { AdKey, FinalScore } from '../features/eval/scoring'

// ---- 後端 shape(backend/app/api/v1/admin_eval.py)----

// 總覽只回總分:逐項明細各社 9 列 × 全校社團數,清單頁一列也用不到
interface AdminEvalClubOut {
  club_id: number
  club_name: string
  attribute: string | null
  year: number
  total: number
}

interface AdjustmentOut {
  id: number
  kind: string // admin_score_override / merit_bonus
  value: Record<string, unknown> // {key, score} 或 {score}
  reason: string
  revoked: boolean
  created_at: string
}

interface AdminEvalDetailOut {
  club_id: number
  club_name: string
  year: number
  total: number
  scores: AdScoreOut[]
  adjustments: AdjustmentOut[]
}

// ---- 前端型別 ----

export interface AdminEvalClub {
  clubId: number
  clubName: string
  attribute: string | null
  year: number
  total: number
}

export interface EvalAdjustment {
  id: number
  kind: 'admin_score_override' | 'merit_bonus'
  key?: AdKey
  score: number
  reason: string
  revoked: boolean
  createdAt: string
}

export interface AdminEvalDetail {
  clubId: number
  clubName: string
  year: number
  total: number
  scores: FinalScore[]
  adjustments: EvalAdjustment[]
  /** 現行表現優良加分(最新未註銷 merit_bonus;無則 0) */
  merit: number
  /** 各項現行調整的原因(tooltip 用;鍵=已調整的 ad key) */
  overrideReasons: Partial<Record<AdKey, string>>
}

// ---- 轉換 ----

const toAdjustment = (a: AdjustmentOut): EvalAdjustment => ({
  id: a.id,
  kind: a.kind as EvalAdjustment['kind'],
  key: typeof a.value.key === 'string' ? (a.value.key as AdKey) : undefined,
  score: typeof a.value.score === 'number' ? a.value.score : 0,
  reason: a.reason,
  revoked: a.revoked,
  createdAt: dayjs(a.created_at).format('YYYY/MM/DD HH:mm'),
})

const toClub = (c: AdminEvalClubOut): AdminEvalClub => ({
  clubId: c.club_id,
  clubName: c.club_name,
  attribute: c.attribute,
  year: c.year,
  total: c.total,
})

const toDetail = (d: AdminEvalDetailOut): AdminEvalDetail => {
  // 後端依 id 降冪回傳:第一筆未註銷者即現行值
  const adjustments = d.adjustments.map(toAdjustment)
  const merit = adjustments.find((a) => a.kind === 'merit_bonus' && !a.revoked)?.score ?? 0
  const overrideReasons: Partial<Record<AdKey, string>> = {}
  for (const a of adjustments) {
    if (a.kind === 'admin_score_override' && !a.revoked && a.key && !(a.key in overrideReasons)) {
      overrideReasons[a.key] = a.reason
    }
  }
  return {
    clubId: d.club_id,
    clubName: d.club_name,
    year: d.year,
    total: d.total,
    scores: toFinalScores(d.scores),
    adjustments,
    merit,
    overrideReasons,
  }
}

// ---- 查詢 ----

const keys = {
  all: ['adminEval'] as const,
  clubs: (page: number) => ['adminEval', 'clubs', page] as const,
  detail: (clubId: number) => ['adminEval', 'detail', clubId] as const,
}

export const adminEvalKeys = keys

export const EVAL_CLUBS_PAGE_SIZE = 20

/** 各社行政分:伺服器端分頁(分數只算這一頁,社團名升冪由後端排) */
export function useAdminEvalClubs(page: number) {
  return useQuery({
    queryKey: keys.clubs(page),
    queryFn: () =>
      apiPaged<AdminEvalClubOut[]>(
        `/admin/eval/clubs${qs({ page, page_size: EVAL_CLUBS_PAGE_SIZE })}`,
      ).then(({ data, total }) => ({ rows: data.map(toClub), total })),
  })
}

export function useAdminEvalDetail(clubId: number | null) {
  return useQuery({
    queryKey: keys.detail(clubId ?? 0),
    enabled: clubId != null,
    // 不留前一社的資料:這份查詢只以 clubId 為鍵,沿用舊值等於把 A 社的分數
    // 攤在 B 社的彈窗上,而彈窗的調整動作照樣可按
    queryFn: () => api<AdminEvalDetailOut>(`/admin/eval/clubs/${clubId}`).then(toDetail),
  })
}

// ---- 調整(override/revert/merit;reason 後端必填)----

export interface OverrideInput {
  key: AdKey
  score: number
  reason: string
}

export interface RevertInput {
  key: AdKey
  reason: string
}

export interface MeritInput {
  score: number
  reason: string
}

export function useAdminEvalMutations(clubId: number | null) {
  const qc = useQueryClient()
  const invalidate = () => void qc.invalidateQueries({ queryKey: keys.all })
  const post = (path: string, body: Record<string, unknown>) =>
    api<{ total: number; scores: AdScoreOut[] }>(`/admin/eval/clubs/${clubId}/${path}`, {
      method: 'POST',
      body: JSON.stringify(body),
    })
  const override = useMutation({
    mutationFn: (v: OverrideInput) => post('override', { key: v.key, score: v.score, reason: v.reason }),
    onSuccess: invalidate,
  })
  const revert = useMutation({
    mutationFn: (v: RevertInput) => post('revert', { key: v.key, reason: v.reason }),
    onSuccess: invalidate,
  })
  const merit = useMutation({
    mutationFn: (v: MeritInput) => post('merit', { score: v.score, reason: v.reason }),
    onSuccess: invalidate,
  })
  return { override, revert, merit }
}

File: backend/app/api/v1/admin_eval.py
"""行政端:行政分審核(逐項手動調整/回到自動/表現優良加分)。

調整全部留痕於 eval_adjustments;「回到自動」=註銷該項現行調整列(revoked_at),
不硬刪,歷次調整可稽核。調整即時反映社團端(查詢時調整值蓋過計算值)。
"""

from typing import Annotated

import sqlalchemy as sa
from fastapi import APIRouter, BackgroundTasks, Depends, Request

from app.api.pagination import Pagination
from app.core.deps import CurrentUser, DbDep, client_ip, require_permission
from app.core.errors import not_found
from app.models import Club, EvalAdjustment
from app.models.enums import AdjustmentKind
from app.schemas.admin import MeritIn, ScoreOverrideIn, ScoreRevertIn
from app.schemas.common import ApiResponse
from app.schemas.eval import AdScoreOut
from app.services import audit, evaluation, notify
from app.services.scoring import apply_overrides, compute_ad_scores, total_of

router = APIRouter(prefix="/admin/eval", tags=["admin"])

EvalAdmin = Annotated[CurrentUser, Depends(require_permission("aeval"))]

_CLUB_AWARD = "club"  # 行政分屬最佳社團獎


async def _club_or_404(db, club_id: int) -> Club:
    club = await db.get(Club, club_id)
    if club is None or not club.is_active:
        raise not_found("找不到社團")
    return club


async def _club_scores(db, club_id: int) -> tuple[list[AdScoreOut], float, int]:
    window = await evaluation.get_eval_window(db)
    scoring_input = await evaluation.gather_scoring_input(db, club_id, window)
    overrides = await evaluation.get_overrides(db, club_id, window.year)
    finals = apply_overrides(compute_ad_scores(scoring_input), overrides)
    return [AdScoreOut(**vars(s)) for s in finals], total_of(finals), window.year


@router.get("/clubs")
async def list_clubs(user: EvalAdmin, db: DbDep, page: Pagination) -> ApiResponse[list[dict]]:
    """各社行政分總覽:只回總分,逐項明細由 /clubs/{id} 供應。

    每個來源各查一次(見 evaluation.gather_scoring_inputs),往返次數不隨社團數成長;
    分數只算這一頁的社團 —— 一頁用不到的 159 社行政分不必每次重算。
    """
    query = sa.select(Club).where(Club.is_active.is_(True)).order_by(Club.name, Club.id)
    total = await db.scalar(sa.select(sa.func.count()).select_from(query.subquery()))
    clubs = (await db.scalars(query.offset(page.offset).limit(page.page_size))).all()
    window = await evaluation.get_eval_window(db)
    club_ids = [c.id for c in clubs]
    inputs = await evaluation.gather_scoring_inputs(db, club_ids, window)
    overrides = await evaluation.get_overrides_by_club(db, club_ids, window.year)
    return ApiResponse(
        data=[
            {
                "club_id": club.id,
                "club_name": club.name,
                "attribute": club.attribute.value if club.attribute else None,
                "year": window.year,
                "total": total_of(
                    apply_overrides(compute_ad_scores(inputs[club.id]), overrides[club.id])
                ),
            }
            for club in clubs
        ],
        meta=page.meta(total or 0),
    )


@router.get("/clubs/{club_id}")
async def club_detail(club_id: int, user: EvalAdmin, db: DbDep) -> ApiResponse[dict]:
    club = await _club_or_404(db, club_id)
    scores, total, year = await _club_scores(db, club.id)
    adjustments = (
        await db.scalars(
            sa.select(EvalAdjustment)
            .where(
                EvalAdjustment.club_id == club.id,
                EvalAdjustment.year == year,
                EvalAdjustment.kind.in_(
                    [AdjustmentKind.ADMIN_SCORE_OVERRIDE, AdjustmentKind.MERIT_BONUS]
                ),
            )
            .order_by(EvalAdjustment.id.desc())
        )
    ).all()
    return ApiResponse(
        data={
            "club_id": club.id,
            "club_name": club.name,
            "year": year,
            "total": total,
            "scores": [s.model_dump() for s in scores],
            "adjustments": [
                {
                    "id": a.id,
                    "kind": a.kind.value,
                    "value": a.value,
                    "reason": a.reason,
                    "revoked": a.revoked_at is not None,
                    "created_at": a.created_at.isoformat(),
                }
                for a in adjustments
            ],
        }
    )


async def _revoke_key(db, club_id: int, year: int, key: str) -> None:
    rows = await db.scalars(
        sa.select(EvalAdjustment).where(
            EvalAdjustment.club_id == club_id,
            EvalAdjustment.year == year,
            EvalAdjustment.kind == AdjustmentKind.ADMIN_SCORE_OVERRIDE,
            EvalAdjustment.revoked_at.is_(None),
        )
    )
    for row in rows:
        if row.value.get("key") == key:
            row.revoked_at = sa.func.now()


@router.post("/clubs/{club_id}/override")
async def override_score(
    club_id: int,
    body: ScoreOverrideIn,
    user: EvalAdmin,
    db: DbDep,
    request: Request,
    background: BackgroundTasks,
) -> ApiResponse[dict]:
    club = await _club_or_404(db, club_id)
    window = await evaluation.get_eval_window(db)
    await evaluation.lock_adjustments(db, club.id)
    await _revoke_key(db, club.id, window.year, body.key)  # 新調整取代舊調整
    db.add(
        EvalAdjustment(
            year=window.year,
            award_id=_CLUB_AWARD,
            club_id=club.id,
            kind=AdjustmentKind.ADMIN_SCORE_OVERRIDE,
            value={"key": body.key, "score": body.score},
            reason=body.reason,
            actor_id=user.id,
        )
    )
    audit.record(
        db,
        action="eval_score_overridden",
        user=user,
        detail=f"club={club.id};{body.key}={body.score}",
        ip=client_ip(request),
    )
    await db.commit()
    background.add_task(
        notify.club_event,
        "alert",
        "行政分手動調整",
        f"{club.name}:{body.key} → {body.score}({body.reason})",
        club.discord_webhook_url,
    )
    scores, total, year = await _club_scores(db, club.id)
    return ApiResponse(data={"total": total, "scores": [s.model_dump() for s in scores]})


@router.post("/clubs/{club_id}/revert")
async def revert_score(
    club_id: int,
    body: ScoreRevertIn,
    user: EvalAdmin,
    db: DbDep,
    request: Request,
    background: BackgroundTasks,
) -> ApiResponse[dict]:
    club = await _club_or_404(db, club_id)
    window = await evaluation.get_eval_window(db)
    await evaluation.lock_adjustments(db, club.id)
    await _revoke_key(db, club.id, window.year, body.key)
    audit.record(
        db,
        action="eval_score_reverted",
        user=user,
        detail=f"club={club.id};key={body.key};reason={body.reason}",
        ip=client_ip(request),
    )
    await db.commit()
    background.add_task(
        notify.club_event,
        "alert",
        "行政分回到自動計算",
        f"{club.name}:{body.key}({body.reason})",
        club.discord_webhook_url,
    )
    scores, total, year = await _club_scores(db, club.id)
    return ApiResponse(data={"total": total, "scores": [s.model_dump() for s in scores]})


@router.post("/clubs/{club_id}/merit")
async def set_merit(
    club_id: int,
    body: MeritIn,
    user: EvalAdmin,
    db: DbDep,
    request: Request,
    background: BackgroundTasks,
) -> ApiResponse[dict]:
    club = await _club_or_404(db, club_id)
    window = await evaluation.get_eval_window(db)
    await evaluation.lock_adjustments(db, club.id)
    # 表現優良加分:最新一筆生效,舊列註銷留痕
    rows = await db.scalars(
        sa.select(EvalAdjustment).where(
            EvalAdjustment.club_id == club.id,
            EvalAdjustment.year == window.year,
            EvalAdjustment.kind == AdjustmentKind.MERIT_BONUS,
            EvalAdjustment.revoked_at.is_(None),
        )
    )
    for row in rows:
        row.revoked_at = sa.func.now()
    db.add(
        EvalAdjustment(
            year=window.year,
            award_id=_CLUB_AWARD,
            club_id=club.id,
            kind=AdjustmentKind.MERIT_BONUS,
            value={"score": body.score},
            reason=body.reason,
            actor_id=user.id,
        )
    )
    audit.record(
        db,
        action="eval_merit_set",
        user=user,
        detail=f"club={club.id};score={body.score}",
        ip=client_ip(request),
    )
    await db.commit()
    background.add_task(
        notify.club_event,
        "alert",
        "表現優良加分登錄",
        f"{club.name}:+{body.score}({body.reason})",
        club.discord_webhook_url,
    )
    scores, total, year = await _club_scores(db, club.id)
    return ApiResponse(data={"total": total, "scores": [s.model_dump() for s in scores]})
